DOCS-911: Improving erasure coding core concepts (#929)
minio/docs: source/operations/concepts/erasure-coding.rst

   :local:
   :depth: 2

.. meta::
   :keywords: erasure coding, healing, availability, resiliency
   :description: Information on MinIO Erasure Coding

> This page provides a simplified overview of MinIO's implementation of Erasure Coding.

Reviewer: Wordsmithing idea:
```suggestion
This page provides an introductory overview of MinIO's implementation of Erasure Coding.
```
If you are new to this, it's not exactly easy to understand no matter how its written. Plus there's another `simplified` below, referring to the diagrams.

Reviewer 2: Ooh, good catch. I'd just drop it

```suggestion
This page provides an overview of MinIO's implementation of Erasure Coding.
```

Reviewer: Reading through the rendered page, would also be useful to have a glossary link for `erasure coding` here. The Availability and Resiliency page linked below isn't particularly novice-friendly.

      MinIO automatically partitions the object into data and parity shards.

An object's :term:`Parity` dictates how many parity shards MinIO creates for an object.
   MinIO uses the :term:`erasure set size <erasure set>` and configured parity for the deployment to determine the number of parity and data shards to create for each object, where:

   :math:`N (ERASURE\ SET\ SIZE) = K (DATA) + M (PARITY)`

   .. figure:: /images/erasure/erasure-coding-shard-ec6.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of an object written with EC:6 to an erasure set with 16 drives

   This deployment has an erasure set size of ``16`` and a parity of ``EC:6``. 
   Based on the formula, MinIO creates ``10`` data shards and ``6`` parity shards for each object.

MinIO can use any parity shard to reconstruct any data shard.
   .. figure:: /images/erasure/erasure-coding-shard-reconstruction.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of MinIO using parity shards to reconstruct data shards

      An object has lost two data shards.
      MinIO uses any two of the available parity shards to reconstruct the data shards.

>       Even a single remaining parity shard would be sufficient to reconstruct both missing data shards.

Reviewer 3: No. The simplest explanation is that you need the same number of shards as your data shards.

In the example above there is 10 data shards. This means to reconstruct an object you need *10* valid shards - that can be any combination of data and parity shards.

>       Even a single remaining parity shard would be sufficient to reconstruct both missing data shards.

Author: Ah - @klauspost to confirm:

For *read operations* we can use any 10 shards, data or parity

For *healing operations* we can use 10 shards, data or parity, to reconstruct all remaining shards

That is, if I have 9 data and 1 parity shard remaining out of 16, I could reconstruct all missing shards.

MinIO requires a minimum of ``K`` Data *or* Parity shards to **read** the object.

>    The value ``K`` constitutes the **read quorum** for the deployment.

Author: So they are the *same* value, except for split-brain where it is K+1. Otherwise they are the same. At least that is my understanding.

>    The value ``K`` constitutes the **read quorum** for the deployment.

Reviewer 2: If there's an exception, they cannot be the _same_ value.

Perhaps line 70-71 could be something like this?

```
MinIO requires a minimum number of shards available to **read** an object.
The shards can be either Data **or** Parity shards.
This value, ``K``, is the **read quorum** for the deployment.
```

You could then use something similar, keeping the ``K`` variable, for the write quorum portion.

   Alternatively, the object can tolerate the loss of no more than ``M`` shards of any time while remaining readable.

   .. figure:: /images/erasure/erasure-coding-shard-read-quorum.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of an object maintaining read quorum with only 10 total shards remaining

      An object has 10 data shards and 6 parity shards for a total of 16 shards.
      It has lost 3 data and 3 parity shards for a total of 6 shards lost and 10 remaining.
      The loss of any further shards would result in losing read quorum.

>    An object that has lost read quorum is effectively "lost" and may be recovered through other means such as :ref:`replication resynchronization <minio-bucket-replication-resynchronize>`.

Reviewer 2: perhaps

`"lost" and cannot be directly recovered, but could possibly recovered by other means ...`

Author: I'll take that.

```suggestion
   An object that has lost read quorum is effectively "lost", but could possibly be recovered through other means such as :ref:`replication resynchronization <minio-bucket-replication-resynchronize>`.
```

MinIO similarly requires successful creation of at least ``K`` data shards to **write** the object.
   The value ``K`` constitutes the **write quorum** for the deployment.
   A deployment could lose write quorum due to having insufficient online drives, insufficient drive space, or network issues between nodes.

   .. figure:: /images/erasure/erasure-coding-shard-write-quorum.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of an object maintaining write quorum with only 10 shards remaining.

      MinIO creates 16 shards for an object with parity of ``EC:6``.
      Due to drive errors, only 10 drives remain online for MinIO to use for write operations.
      This meets write quorum, but the loss of any further drives would result in write failure for newer objects.

   If Parity (``EC:M``) is exactly 1/2 the number of drives in the erasure set, **write quorum** is ``K+1``.
   This extra shard prevents data inconsistency due to "split-brain"-types of failure states

.. _minio-ec-erasure-set:

> Erasure Set and Distribution

Reviewer 2: Consider having this section at the top before the discussion of data and parity shards.
It seems to me that understanding an erasure set is a precursor to understanding how sharding works.

----------------------------

> An **Erasure Set** is a group of drives onto which MinIO can read or write erasure coded objects.

Reviewer 2: ```suggestion
An **Erasure Set** is a group of up to sixteen drives for which MinIO can read or write erasure coded objects.
```

Reviewer 2: Perhaps `a group of between two and sixteen drives`

   .. figure:: /images/erasure/erasure-coding-erasure-set.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of erasure set covering 4 nodes and 16 drives

      The above example deployment consists of 4 nodes with 4 drives each.

>       MinIO initializes with a single erasure set consisting of 16 drives.

Reviewer 2: ```suggestion
      MinIO initializes with a single erasure set consisting of all 16 drives across all four nodes.
```

Erasure set stripe size dictates the maximum possible :ref:`parity <minio-ec-parity>` of the deployment.
   .. figure:: /images/erasure/erasure-coding-possible-parity.svg
      :figwidth: 100%
      :align: center
      :alt: Diagram of possible erasure set parity settings

      The above example deployment has an erasure set of 16 drives. 
      This can support parity between ``EC:0`` and 1/2 the erasure set drives, or ``EC:8``.

>    Deployments with a small number of nodes or drives can supported a limited number of parity drives, to a maximum of 1/2 the stripe size.

Reviewer 2: ```suggestion
   Deployments with a small number of nodes or drives can support a limited number of parity drives, up to a maximum of 1/2 the number of drives in the set.
```

>    Erasure Sets have a minimum size of 2 and a maximum size of 16.

Reviewer 2: I suggested adding this earlier in the discussion. If you take that suggestion, you can probably drop it here.

MinIO automatically calculates the number and size of erasure sets when initializing a :ref:`server pool <minio-intro-server-pool>`.

> The stripe size for a pool is **immutable** after it's initial setup.

Reviewer 2: `immutable` may not be easily understood?

Perhaps
```suggestion
You cannot change the stripe size for a pool after its initial setup.
```

Reviewer 3: Now you use "stripe size", whereas you used "set size" above. IMO we should completely drop using "stripe size" as it is confusing.

   :keywords: erasure coding, healing, availability, resiliency
   :description: Information on MinIO Erasure Coding

MinIO implements Erasure Coding as a core component in providing data redundancy and availability.

> This page provides simplified overview of MinIO Erasure Coding.

Author: Fixed manually

- For information on how MinIO uses Erasure Coding for availability in distributed deployments, see :ref:`minio-availability-resiliency`.

> - For information on MinIO deployment architectures, see :ref:`minio-architecture`.

Author: It's not quite hardware configs either...taking another shot at this.

     - 8
     - 9

Bitrot Protection
-----------------

Silent data corruption or bit rot is a serious problem faced by data drives resulting in data getting corrupted without the user’s knowledge. 

> The corruption of data occurs when the electrical charge on a portion of the drive disperses or changes with no notification to or input from the user.

Reviewer: (Not sure how to suggest changes to more than one line.)

This wording feels like I'm about to be sold a personal care product for my embarrassing data problem. Here's an attempt at a rewrite:

> Bit rot is silent data corruption from random changes at the storage media level.
> For data drives, it is typically the result of decay of the electrical charge or magnetic orientation that represents the data.
> Bit rot can cause subtle errors or corruption with no obvious cause and without warning.

Maybe even provide a link https://en.wikipedia.org/wiki/Data_degradation

Author: I'll take it!

MinIO’s optimized implementation of the :minio-git:`HighwayHash algorithm <highwayhash/blob/master/README.md>` ensures that it captures and heals corrupted objects on the fly. 

> Integrity is ensured from end to end by computing a hash on READ and verifying it on WRITE from the application, across the network, and to the memory or drive. 

Reviewer: I'm having trouble parsing what exactly this is saying. Is it that MinIO computes the hash when data is read, and then verifies it as that data is written (which could follow multiple paths to the ultimate destination).
